src/Rooms.ts

```diff
@@ -57,42 +57,49 @@ export function guessAndSetDMRoom(room: Room, isDirect: boolean): Promise<void>
 /**
  * Marks or unmarks the given room as being as a DM room.
  * @param {string} roomId The ID of the room to modify
- * @param {string} userId The user ID of the desired DM
- room target user or null to un-mark
- this room as a DM room
+ * @param {string | null} userId The user ID of the desired DM room target user or
+ *                        null to un-mark this room as a DM room
  * @returns {object} A promise
  */
-export async function setDMRoom(roomId: string, userId: string): Promise<void> {
+export async function setDMRoom(roomId: string, userId: string | null): Promise<void> {
     if (MatrixClientPeg.get().isGuest()) return;
 
     const mDirectEvent = MatrixClientPeg.get().getAccountData(EventType.Direct);
-    let dmRoomMap = {};
+    const currentContent = mDirectEvent?.getContent() || {};
 
-    if (mDirectEvent !== undefined) dmRoomMap = { ...mDirectEvent.getContent() }; // copy as we will mutate
+    // already marked as DM
+    if (userId && currentContent[userId]?.includes?.(roomId)) return;
+
+    const dmRoomMap = new Map(Object.entries(currentContent));
+    let removed = false;
 
     // remove it from the lists of any others users
     // (it can only be a DM room for one person)
-    for (const thisUserId of Object.keys(dmRoomMap)) {
-        const roomList = dmRoomMap[thisUserId];
+    for (const thisUserId of dmRoomMap.keys()) {
+        const roomList = dmRoomMap.get(thisUserId) || [];
 
         if (thisUserId != userId) {
             const indexOfRoom = roomList.indexOf(roomId);
             if (indexOfRoom > -1) {
                 roomList.splice(indexOfRoom, 1);
+                removed = true;
             }
         }
     }
 
+    // already not marked as DM
+    if (userId === null && !removed) return;
+
     // now add it, if it's not already there
     if (userId) {
-        const roomList = dmRoomMap[userId] || [];
+        const roomList = dmRoomMap.get(userId) || [];
         if (roomList.indexOf(roomId) == -1) {
             roomList.push(roomId);
```

Comment:
seems like we could go further here, and suppress the update if no change is made here.

In other words:

1. `s/removed/modified/`
2. here, set `modified=true`
3. move the early return above to below:

```js
if (!modified) {
    // no need to update the account data
    return;
}
```


I'd actually find that clearer to follow.

Reply:
This function is confusing, because it handles two cases when it should probably better be divided into two functions: `setDMRoom(roomId, userId)` and `unsetDMRoom(roomId)`.

Reply:
Done